### PR TITLE
copy: Set copy buffer size based on total physical memory

### PR DESCRIPTION
**Pull request by AndreiSva**
## Description
This commit implements the memory heuristic mechanism suggested by @bluca to set the size of the copy buffer to different values depending on the total physical memory available on the system.

Here are the value steps this patch introduces:
| Physical memory| Copy buffer size |
| ----------- | ----------- |
| <= 128 MB | 8 KB |
| 129 - 256 MB   | 12 KB | 
| 257 - 512 MB   | 24 KB | 
| 513 - 2048  MB   | 32 KB |
| >= 2049  MB   | 64 KB |

## Rationale
- Allocating large buffers on systems with very little RAM (such as embedded systems) can be slow, inefficient and wasteful. As it stands the userspace buffer used in `copy.c` is one of the largest in systemd. This commit adds a mechanism that adapts the size of the buffer based on the total memory available. This allows for increased performance on systems with a large amount of memory while preserving efficiency on embedded hardware.
- Also see #33380

### CODE REVIEW

File: src/shared/copy.c [hunk 1]
```diff
@@ -40,9 +40,7 @@
 #include "umask-util.h"
 #include "user-util.h"
 #include "xattr-util.h"
-
-/* If we copy via a userspace buffer, size it to 64K */
-#define COPY_BUFFER_SIZE (64U*U64_KB)
+#include "limits-util.h"
```
Comment (bluca) [inline src/shared/copy.c]: Sort alphabetically

File: src/shared/copy.c [hunk 2]
```diff
@@ -53,6 +51,32 @@
...
+        /* If we copy via a userspace buffer, try to find a reasonable size for the copy buffer proportional
+           to the total physical memory of the system  */
+
+        static uint64_t size = 0;
+
+        if (size == 0) {
+
+            uint64_t mem = physical_memory();
+
+            if (mem <= 128U*U64_MB) {
```
Comment (bluca) [inline src/shared/copy.c]: No braces around single line blocks

**Comment by poettering** [comment]
This sounds like premature optimization. 

I have serious doubts that this is really a good thing to do. Can you provide us with any kind of profiling/performance data that shows this actually makes a positive difference? Even on 128M RAM a short-lived allocation of 64K doesn't really matter much.

**Comment by AndreiSva** [comment]
I haven't been able to show a meaningful decrease in page faults or increase of performance in my tests. Part of me thinks this is still a marginal improvement, but it is probably not worth the extra syscall.

**Pull request closed without merging**